### PR TITLE
Panel of Normals

### PR DESCRIPTION
**Pull request by MaxUlysse**
Panel of normals (PON) is needed to filter GATK4 Mutect2 somatic calls, though also useful to annotate somatic calls produced by other software also. This PR focuses on Mutect2, and follows instructions at https://gatkforums.broadinstitute.org/gatk/discussion/11136/how-to-call-somatic-mutations-using-gatk4-mutect2 

## PR checklist
 - [x] This comment contains a description of changes (with reason)
 - [ ] If you've fixed a bug or added code that should be tested, add tests!
 - [x] If necessary, also make a PR on the [nf-core/sarek branch on the nf-core/test-datasets repo]( https://github.com/nf-core/test-datasets/pull/new/nf-core/sarek)
 - [ ] Ensure the test suite passes (`nextflow run . -profile test,docker`).
 - [ ] Make sure your code lints (`nf-core lint .`).
 - [x] Documentation in `docs` is updated
 - [x] `CHANGELOG.md` is updated
 - [x] `README.md` is updated

### CODE REVIEW

File: main.nf [hunk 1]
```diff
@@ -1036,42 +1036,232 @@ pairBam = pairBam.dump(tag:'BAM Somatic Pair')
...
 // STEP GATK MUTECT2
 
 process Mutect2 {
     tag {idSampleTumor + "_vs_" + idSampleNormal + "-" + intervalBed.baseName}
 
     input:
-        set idPatient, idSampleNormal, file(bamNormal), file(baiNormal), idSampleTumor, file(bamTumor), file(baiTumor), file(intervalBed) from pairBamMuTect2
-        set file(genomeFile), file(genomeIndex), file(genomeDict), file(dbsnp), file(dbsnpIndex) from Channel.value([
+        set idPatient, idSampleNormal, file(bamNormal), file(baiNormal), idSampleTumor, file(bamTumor), file(baiTumor), file(intervalBed) from pairBamMutect2
+        set file(genomeFile), file(genomeIndex), file(genomeDict), file(intervals), file(commonSNPs), file(commonSNPsIndex) from Channel.value([
```
Comment (MaxUlysse) [inline main.nf]: replace `commonSNPs` with `germline_ressource` to be consistent with GATK parameters

Comment (MaxUlysse) [reply main.nf]: Need to update these as well:
https://github.com/nf-core/sarek/blob/ca2a0b0e41128f86d24ab683074592ca26b81df8/conf/igenomes.config#L16-L17
https://github.com/nf-core/sarek/blob/ca2a0b0e41128f86d24ab683074592ca26b81df8/conf/igenomes.config#L31-L32
https://github.com/nf-core/sarek/blob/ca2a0b0e41128f86d24ab683074592ca26b81df8/conf/genomes.config#L16-L17
https://github.com/nf-core/sarek/blob/ca2a0b0e41128f86d24ab683074592ca26b81df8/conf/genomes.config#L31-L32
https://github.com/nf-core/sarek/blob/ca2a0b0e41128f86d24ab683074592ca26b81df8/conf/genomes.config#L46-L47

**Comment by MaxUlysse** [comment]
You need to include the `igenomes.config` and `genomes.config` files, and also update the docs with `Mutect2` instead of `MuTect2`

**Comment by MaxUlysse** [comment]
And if you could add some docs about the PONs that would be very nice as well

File: conf/igenomes.config [hunk 2]
```diff
@@ -30,6 +30,8 @@
        bwaIndex         = "${params.igenomes_base}/Homo_sapiens/GATK/GRCh38/Sequence/BWAIndex/Homo_sapiens_assembly38.fasta.64.{alt,amb,ann,bwt,pac,sa}"
        dbsnp            = "${params.igenomes_base}/Homo_sapiens/GATK/GRCh38/Annotation/GATKBundle/dbsnp_146.hg38.vcf.gz"
        dbsnpIndex       = "${params.igenomes_base}/Homo_sapiens/GATK/GRCh38/Annotation/GATKBundle/dbsnp_146.hg38.vcf.gz.tbi"
+       germlineResource      = "${params.igenomes_base}/Homo_sapiens/GATK/GRCh38/Annotation/GermlineResource/GCF_000001405.38.AUTOSOMESXY.COMMON.BIALLELIC.SNPs.with.AF.vcf.gz"
```
Comment (MaxUlysse) [inline conf/igenomes.config]: We don't have that for `GRCh37`?

Comment (MaxUlysse) [reply conf/igenomes.config]: cf #25

File: main.nf [hunk 3]
```diff
@@ -121,6 +121,7 @@ params.multiqc_config = null
 params.noGVCF = null
 params.noStrelkaBP = null
 params.nucleotidesPerSecond = 1000.0
+params.pon = null
```
Comment (MaxUlysse) [inline main.nf]: Can you describe this params in `docs/usage.md` and in the `helpMessage` in `main.nf`

File: conf/genomes.config [hunk 4]
```diff
@@ -25,34 +25,38 @@ params {
...
-      knownIndels      = "${params.genomes_base}/{Mills_and_1000G_gold_standard.indels.hg38,beta/Homo_sapiens_assembly38.known_indels}.vcf.gz"
-      knownIndelsIndex = "${params.genomes_base}/{Mills_and_1000G_gold_standard.indels.hg38,beta/Homo_sapiens_assembly38.known_indels}.vcf.gz.tbi"
-      snpeffDb         = "GRCh38.86"
-      vepCacheVersion  = "95"
+      acLoci                = "${params.genomes_base}/1000G_phase3_GRCh38_maf0.3.loci"
+      acLociGC              = "${params.genomes_base}/1000G_phase3_GRCh38_maf0.3.loci.gc"
+      bwaIndex              = "${params.genomes_base}/Homo_sapiens_assembly38.fasta.64.{alt,amb,ann,bwt,pac,sa}"
+      dbsnp                 = "${params.genomes_base}/dbsnp_146.hg38.vcf.gz"
+      dbsnpIndex            = "${params.genomes_base}/dbsnp_146.hg38.vcf.gz.tbi"
+      germlineResource      = "${params.genomes_base}/GCF_000001405.38.AUTOSOMESXY.COMMON.BIALLELIC.SNPs.with.AF.vcf.gz"
```
Comment (MaxUlysse) [inline conf/genomes.config]: Can you describe this references in `docs/usage.md`, `docs/reference.md` and in the `helpMessage` in `main.nf`

File: docs/output.md [hunk 5]
```diff
@@ -140,16 +140,16 @@ For all samples:
...
 
-For further reading and documentation see the [MuTect2 manual](https://software.broadinstitute.org/gatk/documentation/tooldocs/4.1.2.0/org_broadinstitute_hellbender_tools_walkers_mutect_Mutect2.php).
+For further reading and documentation see the [Mutect2 manual](https://software.broadinstitute.org/gatk/documentation/tooldocs/4.1.2.0/org_broadinstitute_hellbender_tools_walkers_mutect_Mutect2.php).
 
 For a Tumor/Normal pair only:
-**Output directory: `results/VariantCalling/[TUMOR_vs_NORMAL]/MuTect2`**
+**Output directory: `results/VariantCalling/[TUMOR_vs_NORMAL]/Mutect2`**
 
-* `MuTect2_[TUMORSAMPLE]_vs_[NORMALSAMPLE].vcf.gz` and `MuTect2_[TUMORSAMPLE]_vs_[NORMALSAMPLE].vcf.gz.tbi`
+* `Mutect2_[TUMORSAMPLE]_vs_[NORMALSAMPLE].vcf.gz` and `Mutect2_[TUMORSAMPLE]_vs_[NORMALSAMPLE].vcf.gz.tbi`
```
Comment (MaxUlysse) [inline docs/output.md]: Can you describe the new organisation of files here?

File: docs/output.md [hunk 6]
```diff
@@ -140,16 +140,16 @@ For all samples:
...
   * VCF with Tabix index
 
-### MuTect2
+### Mutect2
 
-[GATK MuTect2](https://github.com/broadinstitute/gatk) calls somatic SNVs and indels via local assembly of haplotypes.
+[GATK Mutect2](https://github.com/broadinstitute/gatk) calls somatic SNVs and indels via local assembly of haplotypes.
 
-For further reading and documentation see the [MuTect2 manual](https://software.broadinstitute.org/gatk/documentation/tooldocs/4.1.2.0/org_broadinstitute_hellbender_tools_walkers_mutect_Mutect2.php).
+For further reading and documentation see the [Mutect2 manual](https://software.broadinstitute.org/gatk/documentation/tooldocs/4.1.2.0/org_broadinstitute_hellbender_tools_walkers_mutect_Mutect2.php).
```
Comment (MaxUlysse) [inline docs/output.md]: Can you put some lines about PON, maybe add also the comment that's in the `main.nf` script:
> please make a panel-of-normals, using at least 40 samples
> https://gatkforums.broadinstitute.org/gatk/discussion/11136/how-to-call-somatic-mutations-using-gatk4-mutect2

File: main.nf (hunk 3 above)
Comment (MaxUlysse) [reply main.nf]: And can you print that params somewhere here as well: https://github.com/nf-core/sarek/blob/d3b3b31899ab0822d2486f4f1451ddfc08942ec3/main.nf#L209-L229

File: main.nf [hunk 7]
```diff
@@ -1072,49 +1073,107 @@ pairBam = bamNormal.cross(bamTumor).map {
...
+(mutect2Stats,intervalStatsFiles) = mutect2Stats.into(2)
+mutect2Stats = mutect2Stats.groupTuple(by:[0,1,2])
+
+process MergeMutect2Stats {
+    tag {idSampleTumor + "_vs_" + idSampleNormal}
+
+    publishDir "${params.outdir}/VariantCalling/${idSampleTumor}_vs_${idSampleNormal}/Mutect2", mode: params.publishDirMode
+
+    input:
+        set caller,
```
Comment (MaxUlysse) [inline main.nf]: I don't think you have `caller` here in this channel.

Comment (MaxUlysse) [reply main.nf]: You could also group the two sets into just one with another `groupTuple`

File: main.nf [hunk 8]
```diff
@@ -1174,19 +1236,156 @@ process ConcatVCF {
...
 }
 
+(vcfConcatenated, vcfConcatenatedForFilter) = vcfConcatenated.into(2)
 vcfConcatenated = vcfConcatenated.dump(tag:'VCF')
 
+process PileupSummariesForMutect2 {
+    tag {idSampleTumor + "_vs_" + idSampleNormal + "_" + intervalBed.baseName }
+
+    input:
+        set idPatient, idSampleNormal, file(bamNormal), file(baiNormal), idSampleTumor, file(bamTumor), file(baiTumor), file(intervalBed) from pairBamPileupSummaries
```
Comment (MaxUlysse) [inline main.nf]: I don't think `bamNormal` is needed in this process.

File: main.nf [hunk 9]
```diff
@@ -1174,19 +1236,156 @@ process ConcatVCF {
...
+
+pileupSummaries = pileupSummaries.groupTuple(by:[0,1])
+
+process MergePileupSummaries {
+    tag {idPatient + "_" + idSampleTumor}
+
+    publishDir "${params.outdir}/VariantCalling/${idSampleTumor}/Mutect2", mode: params.publishDirMode
+
+    input:
+        file(genomeDict) from referenceMap.genomeDict
```
Comment (MaxUlysse) [inline main.nf]:
```suggestion
        file(genomeDict) from referenceMap.genomeDict
```

Comment (MaxUlysse) [inline main.nf]:
```suggestion
        file(genomeDict) from Channel.value([referenceMap.genomeDict])
```

File: main.nf [hunk 10]
```diff
@@ -1636,10 +1835,10 @@ controlFreecVizOut.dump(tag:'ControlFreecViz')
 (vcfMantaSomaticSV, vcfMantaDiploidSV) = vcfManta.into(2)
 
 vcfKeep = Channel.empty().mix(
-    vcfConcatenated.map {
-        variantcaller, idPatient, idSample, vcf, tbi ->
-        [variantcaller, idSample, vcf]
-    },
+//    vcfConcatenated.map {
```
Comment (MaxUlysse) [inline main.nf]: Does this need to stay commented out?

File: main.nf (hunk 7 above)
Comment (szilvajuhos) [reply main.nf]: Now there is the missing caller - I would prefer to leave it it as it is as there is a fork later and it is already pretty cluttered.

File: main.nf (hunk 8 above)
Comment (szilvajuhos) [reply main.nf]: That is true, but simpler now to multiply the channels for variant callers. We can dissect the channel and re-assemble if you feel it is bad style.

File: main.nf (hunk 10 above)
Comment (szilvajuhos) [reply main.nf]: nope, deleted

File: main.nf (hunk 9 above)
Comment (szilvajuhos) [reply main.nf]: why to make an extra channel?

File: main.nf (hunk 3 above)
Comment (szilvajuhos) [reply main.nf]: Done.

File: docs/output.md (hunk 5 above)
Comment (szilvajuhos) [reply docs/output.md]: Done

**Comment by apeltzer** [comment]
The error handling code is buggy still (from the current template version unfortunately):

```bash
if (workflow.stats.ignoredCount > 0 && workflow.success) {
      log.info "${c_purple}Warning, pipeline completed, but with errored process(es) ${c_reset}"
      log.info "${c_red}Number of ignored errored process(es) : ${workflow.stats.ignoredCount} ${c_reset}"
      log.info "${c_green}Number of successfully ran process(es) : ${workflow.stats.succeedCount} ${c_reset}"
    }
```

Is the way to do it - can do it via an own PR ...

File: main.nf (hunk 9 above)
Comment (apeltzer) [reply main.nf]: I think the behavior is the same in this case, you'll always get a `ValueChannel` either way: The latter proposed way just states it explicitly, so even someone not being aware of this behavior of nextflow will know this is a `ValueChannel` and not a `QueueChannel` now.

**Pull request merged**